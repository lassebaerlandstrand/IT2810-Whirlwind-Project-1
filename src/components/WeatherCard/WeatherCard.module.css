.city {
  font-size: 1rem;
}

.temperature {
  font-size: 3rem;
  text-align: center;
}

.card {
  background-color: var(--primary-color);
  color: var(--secondary-color);
  border-radius: 2rem;
  padding: 1rem 3rem;
  margin: 1rem auto;
  box-shadow: 0 0 1rem rgba(0, 0, 0, 0.5);
  max-width: fit-content;
  text-align: center;
}

svg {
  width: 100%;
  height: 100%;
  fill: var(--secondary-color);
}

p {
  margin: 0;
}
