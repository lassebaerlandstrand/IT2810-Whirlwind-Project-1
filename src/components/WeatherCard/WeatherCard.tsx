import {
  IconCloudFilled,
  IconCloudRain,
  IconHeart,
  IconHeartFilled,
  IconRotateClockwise,
  IconSunFilled,
} from '@tabler/icons-react';
import React from 'react';
import { useFavorites } from '../../contexts/FavoritesFunctions';
import { Location, WeatherInfo } from '../../types/api-types';
import styles from './WeatherCard.module.css';

const determineWeather = (data: WeatherInfo) => {
  if (data.precipitation_amount > 0) {
    return [<IconCloudRain className={styles.svg} aria-hidden="true" />, 'Wet'];
  }
  if (data.cloud_area_fraction > 0.5) {
    return [<IconCloudFilled className={styles.svg} aria-hidden="true" />, 'Cloudy'];
  }
  return [<IconSunFilled className={styles.svg} aria-hidden="true" />, 'Clear'];
};

interface WeatherCardProps {
  location: Location;
  data: WeatherInfo | null;
}

const WeatherCard: React.FC<WeatherCardProps> = ({ location, data }) => {
  const { favorites, toggleFavorite } = useFavorites();

  const isFavorite = favorites.some(
    (fav) => fav.city_name === location.city_name && fav.country_name === location.country_name,
  );

  const handleFavoriteClick = (e: React.MouseEvent) => {
    e.stopPropagation();
    toggleFavorite(location);
  };

  return (
    <div className={`${styles.card} ${data && data.air_temperature > 0 ? styles.warmCard : styles.coldCard}`}>
      <button
        className={styles.favoriteIcon}
        onClick={handleFavoriteClick}
        aria-pressed={isFavorite}
        aria-label={
          isFavorite ? `Remove ${location.city_name} from favorites` : `Add ${location.city_name} to favorites`
        }
      >
        {isFavorite ? <IconHeartFilled aria-hidden="true" /> : <IconHeart aria-hidden="true" />}
      </button>

      {data ? (
        determineWeather(data)[0]
      ) : (
        <IconRotateClockwise className={`${styles.rotating} ${styles.svg}`} aria-hidden="true" />
      )}

      <div className={styles.informationContainer}>
        <div className={styles.locationContainer}>
          <h3>{location.city_name}</h3>
          <span className={styles.commaSeparator} aria-hidden="true">
            ,{' '}
          </span>
          <h4>{location.country_name}</h4>
        </div>
        <p
          className={`${styles.temperature} ${data && data.air_temperature > 0 ? styles.warmTemp : styles.coldTemp}`}
          aria-label={`Temperature in ${location.city_name}: ${data ? data.air_temperature : 'loading'}`}
        >
          {data ? data.air_temperature : '---'}Â°C
        </p>
        <p>{data ? determineWeather(data)[1] : 'Loading...'}</p>
      </div>
    </div>
  );
};

export default WeatherCard;
