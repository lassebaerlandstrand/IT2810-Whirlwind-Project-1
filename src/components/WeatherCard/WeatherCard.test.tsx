import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import LOCATIONS from '../../utils/locations';
import WeatherCard from './WeatherCard';

describe('WeatherCard', () => {
  it('renders loading state', () => {
    render(<WeatherCard location={LOCATIONS[0]} data={null} />);

    // Check if loading spinner is displayed
    expect(screen.getByText('Loading...')).toBeInTheDocument();
    expect(screen.getByText('New York')).toBeInTheDocument();
    expect(screen.getByText('---째C')).toBeInTheDocument();
  });

  it('renders clear weather', () => {
    render(
      <WeatherCard
        location={LOCATIONS[0]}
        data={{
          air_pressure_at_sea_level: 0,
          air_temperature: 25,
          cloud_area_fraction: 0.3,
          relative_humidity: 0,
          wind_from_direction: 0,
          wind_speed: 0,
          precipitation_amount: 0,
        }}
      />,
    );

    // Check if proper text is displayed
    expect(screen.getByText('Clear')).toBeInTheDocument();
    expect(screen.getByText('25째C')).toBeInTheDocument();
  });

  it('renders cloudy weather', () => {
    render(
      <WeatherCard
        location={LOCATIONS[0]}
        data={{
          air_pressure_at_sea_level: 0,
          air_temperature: 15,
          cloud_area_fraction: 0.7,
          relative_humidity: 0,
          wind_from_direction: 0,
          wind_speed: 0,
          precipitation_amount: 0,
        }}
      />,
    );

    // Check if cloudy icon and text are displayed
    expect(screen.getByText('Cloudy')).toBeInTheDocument();
    expect(screen.getByText('15째C')).toBeInTheDocument();
  });

  it('renders wet weather', () => {
    render(
      <WeatherCard
        location={LOCATIONS[0]}
        data={{
          air_pressure_at_sea_level: 0,
          air_temperature: 10,
          cloud_area_fraction: 0.8,
          relative_humidity: 0,
          wind_from_direction: 0,
          wind_speed: 0,
          precipitation_amount: 0.5,
        }}
      />,
    );

    // Check if proper text are displayed
    expect(screen.getByText('Wet')).toBeInTheDocument();
    expect(screen.getByText('10째C')).toBeInTheDocument();
  });

  it('handles API error', () => {
    render(<WeatherCard location={LOCATIONS[0]} data={null} />);
    // Check if loading state is still displayed due to error
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <WeatherCard
        location={LOCATIONS[0]}
        data={{
          air_pressure_at_sea_level: 0,
          air_temperature: 25,
          cloud_area_fraction: 0.3,
          relative_humidity: 0,
          wind_from_direction: 0,
          wind_speed: 0,
          precipitation_amount: 0,
        }}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
