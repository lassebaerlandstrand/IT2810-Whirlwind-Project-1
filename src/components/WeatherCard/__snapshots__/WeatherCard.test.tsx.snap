// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeatherCard > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="_card_8ab24e"
  >
    <svg
      class="tabler-icon tabler-icon-sun-filled _svg_8ab24e"
      fill="currentColor"
      height="24"
      stroke="none"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 19a1 1 0 0 1 .993 .883l.007 .117v1a1 1 0 0 1 -1.993 .117l-.007 -.117v-1a1 1 0 0 1 1 -1z"
      />
      <path
        d="M18.313 16.91l.094 .083l.7 .7a1 1 0 0 1 -1.32 1.497l-.094 -.083l-.7 -.7a1 1 0 0 1 1.218 -1.567l.102 .07z"
      />
      <path
        d="M7.007 16.993a1 1 0 0 1 .083 1.32l-.083 .094l-.7 .7a1 1 0 0 1 -1.497 -1.32l.083 -.094l.7 -.7a1 1 0 0 1 1.414 0z"
      />
      <path
        d="M4 11a1 1 0 0 1 .117 1.993l-.117 .007h-1a1 1 0 0 1 -.117 -1.993l.117 -.007h1z"
      />
      <path
        d="M21 11a1 1 0 0 1 .117 1.993l-.117 .007h-1a1 1 0 0 1 -.117 -1.993l.117 -.007h1z"
      />
      <path
        d="M6.213 4.81l.094 .083l.7 .7a1 1 0 0 1 -1.32 1.497l-.094 -.083l-.7 -.7a1 1 0 0 1 1.217 -1.567l.102 .07z"
      />
      <path
        d="M19.107 4.893a1 1 0 0 1 .083 1.32l-.083 .094l-.7 .7a1 1 0 0 1 -1.497 -1.32l.083 -.094l.7 -.7a1 1 0 0 1 1.414 0z"
      />
      <path
        d="M12 2a1 1 0 0 1 .993 .883l.007 .117v1a1 1 0 0 1 -1.993 .117l-.007 -.117v-1a1 1 0 0 1 1 -1z"
      />
      <path
        d="M12 7a5 5 0 1 1 -4.995 5.217l-.005 -.217l.005 -.217a5 5 0 0 1 4.995 -4.783z"
      />
    </svg>
    <div
      class="_informationContainer_8ab24e"
    >
      <div
        class="_locationContainer_8ab24e"
      >
        <h3>
          New York
        </h3>
        <span
          class="_commaSeparator_8ab24e"
        >
          , 
        </span>
        <h4>
          USA
        </h4>
      </div>
      <p
        class="_temperature_8ab24e"
      >
        25Â°C
      </p>
      <p>
        Sunny
      </p>
    </div>
  </div>
</DocumentFragment>
`;
